src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java

```diff
@@ -381,7 +381,7 @@ public synchronized void onNodesRemoved(long epoch, Topology current, Set<Node.I
     private long[] nonCompletedEpochsBefore(long max)
```

Comment:
I don't think it is safe for this method to be called from an unsynchronised block without itself being synchronised

Reply:
there are 3 things that matter:

1) `epochs.minEpoch()`
2) `epochs.maxEpoch()`
3) `snapshot.syncStatus`

first 2 share the epoch lock, and `maxEpoch` could grow between calls to `minEpoch`, but for this context that seems fine.

That leaves `snapshot.syncStatus`, which does need the `AccordConfigurationService.this` lock for visibility...

To limit the locks as much as possible I could lock `getEpochSnapshot` or I can just make syncStatus volatile... given that this is the only code path that isn't accessed with a lock and its via `getEpochSnapshot` which is only called here, I went with sync on `getEpochSnapshot`.

A 3rd option is just accept that its stale... by the time this method exits the status might flip to complete, so we double add the node to `remote_sync_complete` in the topologies table, since this is a set it will dedup for us.

Reply:
We can prune, so min epoch can move also.